lib/oauth2-client.js

```diff
@@ -0,0 +1,131 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+
+import database from 'bedrock-mongodb';
+import * as bedrock from 'bedrock';
+import {promisify} from 'util';
+import {httpClient} from '@digitalbazaar/http-client';
+const {util: {BedrockError}} = bedrock;
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await promisify(database.openCollections)(['access-tokens']);
+
+  // shard based on `siteId`
+  await promisify(database.createIndexes)([{
+    // prevent duplicate events from being stored
+    collection: 'access-tokens',
+    fields: {'token.token': 1},
+    options: {unique: true, background: false}
+  }, {
+    // automatically expire events with an `expires` date field
+    collection: 'access-tokens',
+    fields: {'token.expires': 1},
+    options: {
+      partialFilterExpression: {'token.expires': {$exists: true}},
+      unique: false,
+      background: false,
+      expireAfterSeconds: 0
+    }
+  }]);
+});
+
+export class OAuth2Client {
+  constructor({
+    client_id, client_secret, url, grant_type, scope, httpsAgent
+  } = {}) {
+    this.client_id = client_id;
+    this.client_secret = client_secret;
+    this.url = url;
+    this.grant_type = grant_type;
+    this.scope = scope;
+    this.httpsAgent = httpsAgent;
+  }
+  async getToken() {
+    const {client_id, client_secret, url, grant_type, scope, httpsAgent} = this;
+    let token;
+    token = await checkDatabase({client_id});
+    if(!token || !token.access_token) {
+      token = await getNewAccessToken({
+        client_id, client_secret, url, grant_type, scope, httpsAgent
+      });
+      await insertToken({token});
+    }
+    return token.access_token;
+  }
+}
+
+async function getNewAccessToken({
```

Comment:
It'd be great to have jsdocs, here and on other functions.

Reply:
Added!